### PR TITLE
[FX][export] DCE pass, check schema for node impurity

### PR DESCRIPTION
**Pull request by yushangdi**
Change the default DCE pass to check node schema for impure nodes.

### CODE REVIEW

**Comment by pianpwk** [comment]
@angelayi sorry to ask, would you have any thoughts on who to ping for this change? The `is_impure()` code is 3 years old and hasn't changed since.

**Comment by yushangdi** [comment]
> @angelayi sorry to ask, would you have any thoughts on who to ping for this change? The `is_impure()` code is 3 years old and hasn't changed since.

we should definitely ask if this is ok to change. If we don't want to change it here, alternatively, I can write a new DCE pass with this change and use it for export only.

**Comment by yushangdi** [comment]
imported to check if it breaks anything internally.

**Review by angelayi — DISMISSED** [review]

**Comment by JacobSzwejbka** [comment]
I wouldnt hesitate to change things just because they are old. Half the old things were written with little to no design and half of them were designed in a context thats 3 years old. As long as tests pass it should be fine, and if they aren't thats the fault of whoever owns something reliant on behavior that isnt tested.

**Comment by yushangdi** [comment]
@angelayi There are some tests failing, and the reason is that `_ops.aten._native_batch_norm_legit.default` shouldn't be marked as impure, even though the schema says some inputs are mutable (running mean and running var). Do you know if there's anything special with the batch_norm operator? One simple thing we can do it whitelist it and exclude it from being marked as impure, but I think it indeed mutates things?

cc @bdhirsh

File: torch/fx/node.py [hunk 1]
```diff
@@ -648,9 +645,18 @@ def is_impure(self):
...
+        # than 1 users (all arguments have `self` as a user).
         if self.op == "call_function":
-            return self.target in _side_effectful_functions
+            schema = getattr(self.target, "_schema", None)
+            schema_mutable = False
+            if schema is not None and schema.is_mutable:
+                for (idx, arg) in enumerate(schema.arguments):
+                    if arg.alias_info is not None and arg.alias_info.is_write:  # arg is mutable
+                        if len(self.args[idx].users) > 1:  # has other users except self
+                            schema_mutable = True
```
Comment (angelayi) [inline torch/fx/node.py:658]: what about kwargs?

Comment (yushangdi) [reply torch/fx/node.py]: added kwargs now.

Comment (yushangdi) [reply torch/fx/node.py]: oh wait, I think there's some issues with the kwargs implementation. I'll fix it and let you know.

Comment (yushangdi) [reply torch/fx/node.py]: should be fixed now!

**Comment by yushangdi** [comment]
@angelayi  I think the current approach might not be correct in general. The problem is that even if a mutable argument has no other explicit users in the graph, we cannot safely say it won't have any side efffects. For example, the mutable argument might be a `placeholder` or `get_attr` (the user is outside of the graph), or it's a `x = y.view()` operator, so though the `x` node itself does not have a user,  `y` can have users. The problem is that I'm not sure if there's a comprehensive list of operators like `view`. 

I think the safest thing to do is to write another DCE pass for non-functional graphs, and be conservative. We check if `schema.is_mutable== True` and if so, we conservatively decide to not remove it. We use this new DCE only for export so we don't break other people's tests.

WDYT?

**Review by angelayi — APPROVED** [review]

**Review by jgong5 — APPROVED** [review]

**Pull request closed without merging**